lib/handlers/router-entities/route-caching/model/cached-routes-strategy.ts

```diff
@@ -1,32 +1,70 @@
-import { Currency, CurrencyAmount } from '@uniswap/sdk-core'
-import { CachedRoutesParameters } from './cached-routes-parameters'
+import { Currency, CurrencyAmount, TradeType } from '@uniswap/sdk-core'
+import { CachedRoutesBucket } from './cached-routes-bucket'
+import { CacheMode, ChainId } from '@uniswap/smart-order-router'
+
+interface CachedRoutesStrategyArgs {
+  pair: string
+  tradeType: TradeType
+  chainId: ChainId
+  buckets: CachedRoutesBucket[]
+}
 
 /**
  * Models out the strategy for categorizing cached routes into buckets by amount traded
  */
 export class CachedRoutesStrategy {
-  private cachingParameters: Map<number, CachedRoutesParameters>
+  readonly pair: string
+  readonly _tradeType: TradeType
+  readonly chainId: ChainId
+  readonly willTapcompare: boolean
   private buckets: number[]
+  private bucketsMap: Map<number, CachedRoutesBucket>
 
   /**
-   * The constructor receives an array of `CachedRoutesParameters`, the extracts and sorts the buckets,
-   * and creates a Map<bucket, CachedRouteParameters>
-   * @param cachedRoutesParameters
+   * @param pair
+   * @param tradeType
+   * @param chainId
+   * @param buckets
    */
-  constructor(cachedRoutesParameters: CachedRoutesParameters[]) {
+  constructor({ pair, tradeType, chainId, buckets }: CachedRoutesStrategyArgs) {
+    this.pair = pair
+    this._tradeType = tradeType
+    this.chainId = chainId
+
+    // Used for deciding to show metrics in the dashboard related to Tapcompare
+    this.willTapcompare = buckets.find((bucket) => bucket.cacheMode == CacheMode.Tapcompare) != undefined
+
     // It is important that we sort the buckets in ascendant order for the algorithm to work correctly.
     // For a strange reason the `.sort()` function was comparing the number as strings, so I had to pass a compareFn.
-    this.buckets = cachedRoutesParameters.map((params) => params.bucket).sort((a, b) => a - b)
+    this.buckets = buckets.map((params) => params.bucket).sort((a, b) => a - b)
 
     // Create a Map<bucket, CachedRouteParameters> for easy lookup once we find a bucket.
-    this.cachingParameters = new Map(cachedRoutesParameters.map((params) => [params.bucket, params]))
+    this.bucketsMap = new Map(buckets.map((params) => [params.bucket, params]))
+  }
+
+  public get tradeType(): string {
+    return this._tradeType == TradeType.EXACT_INPUT ? 'ExactIn' : 'ExactOut'
+  }
+
+  public readablePairTradeTypeChainId(): string {
+    return `${this.pair.toUpperCase()}/${this.tradeType}/${this.chainId}`
+  }
+
+  public bucketPairs(): [number, number][] {
+    const firstBucket: [number, number][] = [[0, this.buckets[0]]]
+    const middleBuckets: [number, number][] = this.buckets
+      .slice(0, -1)
+      .map((bucket, i): [number, number] => [bucket, this.buckets[i + 1]!])
```

Comment:
If `buckets` is of length 1 , `this.buckets[i + 1]!` will be out of range right? Not sure if the result of slice(0, -1) would have one element or no elements in this case though.

Reply:
great catch, I fixed this, and also handled the case of 0 buckets.